### PR TITLE
Flesh out bigquery API

### PR DESCRIPTION
**Pull request by tseaver**
- Add missing top-level convenience imports
- Add missing `Client.list_datasets` and `Dataset.list_tables` methods.

### CODE REVIEW

File: gcloud/bigquery/dataset.py [hunk 1]
```diff
@@ -190,6 +190,25 @@ def location(self, value):
...
+        :param resource: dataset resource representation returned from the API
+
+        :type client: :class:`gcloud.bigquery.client.Client`
+        :param client: Client which holds credentials and project
+                       configuration for the dataset.
+
+        :rtype: :class:`gcloud.bigquery.dataset.Dataset`
+        :returns: Dataset parsed from ``resource``.
+        """
+        name = resource['datasetReference']['datasetId']
```
Comment (dhermes) [inline gcloud/bigquery/dataset.py]: Not worried about `KeyError`s? (Am I sounding like a broken record yet?)

File: gcloud/bigquery/client.py [hunk 2]
```diff
@@ -43,6 +43,49 @@ class Client(JSONClient):
...
 
+    def list_datasets(self, include_all=False, max_results=None,
+                      page_token=None):
+        """List datasets for the project associated with this client.
+
+        See:
+        https://cloud.google.com/pubsub/reference/rest/v1beta2/projects/datasets/list
+
+        :type include_all: boolean
+        :param include_all: Should results include hidden datasets?
```
Comment (dhermes) [inline gcloud/bigquery/client.py]: Question mark?

File: gcloud/bigquery/client.py [hunk 3]
```diff
@@ -43,6 +43,49 @@ class Client(JSONClient):
...
+                            passed, defaults to a value set by the API.
+
+        :type page_token: string
+        :param page_token: opaque marker for the next "page" of datasets. If
+                           not passed, the API will return the first page of
+                           datasets.
+
+        :rtype: tuple, (list, str)
+        :returns: list of :class:`gcloud.pubsub.dataset.Dataset`, plus a
+                  "next page token" string:  if not None, indicates that
```
Comment (dhermes) [inline gcloud/bigquery/client.py]: To clarify, say "if the token is not None"

File: gcloud/bigquery/table.py [hunk 4]
```diff
@@ -298,6 +298,24 @@ def view_query(self):
...
+        :type resource: dict
+        :param resource: table resource representation returned from the API
+
+        :type dataset: :class:`gcloud.bigquery.dataset.Dataset`
+        :param dataset: The dataset containing the table.
+
+        :rtype: :class:`gcloud.bigquery.table.Table`
+        :returns: Table parsed from ``resource``.
+        """
+        table_name = resource['tableReference']['tableId']
```
Comment (dhermes) [inline gcloud/bigquery/table.py]: Ditto here on `KeyError`s

File: gcloud/bigquery/client.py [hunk 5]
```diff
@@ -43,6 +43,49 @@ class Client(JSONClient):
 
     _connection_class = Connection
 
+    def list_datasets(self, include_all=False, max_results=None,
+                      page_token=None):
+        """List datasets for the project associated with this client.
+
+        See:
+        https://cloud.google.com/pubsub/reference/rest/v1beta2/projects/datasets/list
+
```
Comment (dhermes) [inline gcloud/bigquery/client.py]: Probably meant `bigquery` not `pubsub`?

File: gcloud/bigquery/dataset.py [hunk 6]
```diff
@@ -357,6 +376,43 @@ def delete(self, client=None):
         client = self._require_client(client)
         client.connection.api_request(method='DELETE', path=self.path)
 
+    def list_tables(self, max_results=None, page_token=None):
+        """List tables for the project associated with this client.
+
+        See:
+        https://cloud.google.com/pubsub/reference/rest/v1beta2/projects/tables/list
+
```
Comment (dhermes) [inline gcloud/bigquery/dataset.py]: Pubsub

File: gcloud/bigquery/dataset.py [hunk 7]
```diff
@@ -357,6 +376,43 @@ def delete(self, client=None):
...
+        :param max_results: maximum number of tables to return, If not
+                            passed, defaults to a value set by the API.
+
+        :type page_token: string
+        :param page_token: opaque marker for the next "page" of datasets. If
+                           not passed, the API will return the first page of
+                           datasets.
+
+        :rtype: tuple, (list, str)
+        :returns: list of :class:`gcloud.pubsub.table.Table`, plus a
```
Comment (dhermes) [inline gcloud/bigquery/dataset.py]: Pubsub

File: gcloud/bigquery/client.py [hunk 8]
```diff
@@ -43,6 +43,49 @@ class Client(JSONClient):
...
+        :param max_results: maximum number of datasets to return, If not
+                            passed, defaults to a value set by the API.
+
+        :type page_token: string
+        :param page_token: opaque marker for the next "page" of datasets. If
+                           not passed, the API will return the first page of
+                           datasets.
+
+        :rtype: tuple, (list, str)
+        :returns: list of :class:`gcloud.pubsub.dataset.Dataset`, plus a
```
Comment (dhermes) [inline gcloud/bigquery/client.py]: Pubsub

**Comment by dhermes** [comment]
Really easy to review! Thanks.

Only a few hangups, `pubsub` in docstrings, `KeyError` questions in factories and use of `?` in a docstring.

File: gcloud/bigquery/dataset.py (hunk 1 above)
Comment (tseaver) [inline gcloud/bigquery/dataset.py]: These exist to parse server-generate resources:  the `datasetReference` structure is *always* there (unlike some of the other properties).  Maybe we should rename them to `_from_api_repr` to indicate that they are private.

File: gcloud/bigquery/client.py (hunk 2 above)
Comment (tseaver) [inline gcloud/bigquery/client.py]: The value is the answer to the question.

File: gcloud/bigquery/client.py (hunk 8 above)
Comment (tseaver) [inline gcloud/bigquery/client.py]: 42b5e260d3848c25549defd54cf49c5a0cdf7fa7

File: gcloud/bigquery/dataset.py (hunk 7 above)
Comment (tseaver) [inline gcloud/bigquery/dataset.py]: 42b5e260d3848c25549defd54cf49c5a0cdf7fa7

File: gcloud/bigquery/dataset.py (hunk 6 above)
Comment (tseaver) [inline gcloud/bigquery/dataset.py]: 42b5e260d3848c25549defd54cf49c5a0cdf7fa7

File: gcloud/bigquery/client.py (hunk 5 above)
Comment (tseaver) [inline gcloud/bigquery/client.py]: 42b5e260d3848c25549defd54cf49c5a0cdf7fa7

File: gcloud/bigquery/table.py (hunk 4 above)
Comment (tseaver) [inline gcloud/bigquery/table.py]: Similarly, the `tableReference` element is always present in resources generated by the back-end.

File: gcloud/bigquery/client.py (hunk 3 above)
Comment (tseaver) [inline gcloud/bigquery/client.py]: 8b4d6f314b07ad67f5737182e717b0ede424ed97

File: gcloud/bigquery/client.py (hunk 2 above)
Comment (tseaver) [inline gcloud/bigquery/client.py]: 531450789ce102a903e80c236c3c65c70a47a1bc

File: gcloud/bigquery/dataset.py (hunk 1 above)
Comment (dhermes) [inline gcloud/bigquery/dataset.py]: Yes I realize, but sometimes the server behaves strangely. E.g. when we had the pub/sub issue with receivedMessages in #901

**Comment by tseaver** [comment]
@dhermes I think everything is resolved, except the question about `KeyErrors` for missing `datasetReference` / `tableReference` entries:  my take is that we should go forward as is:

- We have no evidence that the back-end fails to set those values.
- Without them, we have no way to construct a `Dataset` / `Table` which could be used to make API requests (without a `name`), which means the `KeyError` is the earliest possible detection for the useless resource.

File: gcloud/bigquery/client.py [hunk 9]
```diff
@@ -43,6 +43,49 @@ class Client(JSONClient):
 
     _connection_class = Connection
 
+    def list_datasets(self, include_all=False, max_results=None,
+                      page_token=None):
+        """List datasets for the project associated with this client.
+
+        See:
+        https://cloud.google.com/bigquery/reference/rest/v1beta2/projects/datasets/list
```
Comment (dhermes) [inline gcloud/bigquery/client.py]: This link doesn't actually work.

File: gcloud/bigquery/dataset.py [hunk 10]
```diff
@@ -357,6 +376,43 @@ def delete(self, client=None):
         client = self._require_client(client)
         client.connection.api_request(method='DELETE', path=self.path)
 
+    def list_tables(self, max_results=None, page_token=None):
+        """List tables for the project associated with this client.
+
+        See:
+        https://cloud.google.com/bigquery/reference/rest/v1beta2/projects/tables/list
```
Comment (dhermes) [inline gcloud/bigquery/dataset.py]: This link doesn't actually work.

File: gcloud/bigquery/client.py [hunk 11]
```diff
@@ -43,6 +43,49 @@ class Client(JSONClient):
...
+                            passed, defaults to a value set by the API.
+
+        :type page_token: string
+        :param page_token: opaque marker for the next "page" of datasets. If
+                           not passed, the API will return the first page of
+                           datasets.
+
+        :rtype: tuple, (list, str)
+        :returns: list of :class:`gcloud.bigquery.dataset.Dataset`, plus a
+                  "next page token" string:  if the toke is not None,
```
Comment (dhermes) [inline gcloud/bigquery/client.py]: s/toke/token

**Comment by dhermes** [comment]
I agree that we can be mostly secure in the belief that the backend will send good data and that without those keys, we can't do anything anyhow.

I just wanted to discuss the possibility that we would provide a more specific error message than `KeyError: 'datasetReference'` or `KeyError: 'datasetId'`.

File: gcloud/bigquery/client.py (hunk 11 above)
Comment (tseaver) [inline gcloud/bigquery/client.py]: 2cd2cd2

File: gcloud/bigquery/dataset.py (hunk 10 above)
Comment (tseaver) [inline gcloud/bigquery/dataset.py]: 9a73871

File: gcloud/bigquery/client.py (hunk 9 above)
Comment (tseaver) [inline gcloud/bigquery/client.py]: 9a73871

**Comment by tseaver** [comment]
FWIW, I'd generally prefer not to re-wrap exceptions (losing the original traceback really damages debuggability, for instance).

**Comment by dhermes** [comment]
If we raise one line after the `KeyError` would have occurred, how does that damage debuggability? If the method fails and the user sees an error in a method they've never heard of, isn't that also pretty low quality debuggability?

**Comment by tseaver** [comment]
Hiding the key error *is* the harm I'm talking about:  the user won't be better able to debug some other error more easily than "that key isn't in the resource as expected".

**Comment by dhermes** [comment]
Maybe we are thinking of different things.

This is what I have in mind:

```python
if ('datasetReference' not in resource or
    'datasetId' not in resource['datasetReference']):
  raise KeyError('The resource returned from the server did not contain '
                 'the the necessary information to create a Dataset '
                 'object. The resource needs to contain a dictionary value '
                 'at the datasetReference key and within that dictionary '
                 'needs the datasetId.')
name = resource['datasetReference']['datasetId']
```

What harm does this cause? I'm just suggesting we provide more information than what a `KeyError` would on its own.

**Comment by tseaver** [comment]
When a Python programmer sees a traceback for a `KeyError` where the bottommost line is:

```python
name = resource['datasetReference']['datasetId']
```

doesn't she already know the same information you typed into that waaay-long error message?  If debugging it, she will still dump out the contents of `resource` and try to figure out why those keys are missing.

**Comment by dhermes** [comment]
Yes that's the root of my original question.

Do we want that Python programmer to just see that `KeyError` and try to figure out what `resource` was and where it came from, or do we want to give them more information which will explain why the error occurred. I was leaning towards the latter since it is not a method users would ever invoke, hence their knowledge of the inputs and failure modes would be minimal.

**Comment by tseaver** [comment]
@dhermes 34a747c adds a check such as you suggested.

**Comment by dhermes** [comment]
LGTM

**Pull request merged**